src/v/utils/container_utils.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+template<typename Container>
+concept IterableContainer
+  = requires(Container c, typename Container::iterator i) {
+    { c.begin() } -> std::same_as<typename Container::iterator>;
+    { c.end() } -> std::same_as<typename Container::iterator>;
+    { ++i } -> std::convertible_to<typename Container::iterator>;
+};
+
+template<typename Container>
+concept RangeErasableContainer = requires(
+  Container c, typename Container::iterator i, typename Container::iterator j) {
+    { c.erase(i, j) } -> std::same_as<typename Container::iterator>;
+};
+
+template<typename Container>
+concept SizableContainer = requires(Container c) {
+    { c.size() } -> std::convertible_to<size_t>;
+};
+
+template<typename Container>
+concept ClearableContainer = requires(Container c) {
+    c.clear();
+};
+
+template<typename Container>
+concept AsyncClearRequirements
+  = IterableContainer<Container> && RangeErasableContainer<
+    Container> && SizableContainer<Container> && ClearableContainer<Container>;
+
+/// Clears the given container while yielding to seastar after every
+/// element is erased. This is used to avoid reactor stall warnings
+/// when deleting containers with a large number of objects.
+///
+/// \param c a container to be erased.
+template<typename Container>
+ss::future<>
+async_clear(Container& c, const size_t threshold_size = 1000) requires
+  AsyncClearRequirements<Container> {
```

Comment:
this concept work is cool. it's awesome to have a chance to use them!

just a few thoughts here i had in passing:

i'm wondering here if it would make sense to specialize on the exact, narrow use cases that exist today, and then expand the scope of where async_clear is applied as we acquire more use cases over time.

the rationale is similar to what john mentioned below: it seems like the end game here is that async_clear is specialized to the container implementation (not just the interface) and the element type (e.g. primitive type vs expensive destructor). i think this abstraction issue is already observable here through your preemptive use of the `threshold_size` parameter.

for instance if i have a std::vector<int> what should i do? i could clear it knowing that there is no point is spreading the cost of a destructor call out, because the types are primitive:

```
template<typename T>
requires std::is_trivially_copyable<T>
async_clear(std::vector<T> c) { c.clear(); }
```

but the strategy would be different if we had an instance of our internal `fragmented_vector<T>` type which chunks the storage, and still different perhaps depending on the cost of the destructor. ultimately we only have a handful of container types in the tree.

Reply:
Specializing per container might be the best idea right now. Beyond that there seems to be an [issue](https://github.com/llvm/llvm-project/issues/49689) with LLVM currently when compiling templated coroutines  that is causing the build to fail on the debug build. I'll go ahead and remove template and specialize it to `absl::flat_hash_map`

Reply:
oh wow. nice find with that llvm issue. me and @BenPope and @graphcareful have most recently hit this. i see that one of the maintainers of seastar even has a backport fix for clang-14 and we just upgraded. very nice to see it getting fixed!

my approach had been to avoid coroutines when mixing `if constexpr` and other advanced template stuff.